internal/pkg/docker/dockerfile/lex.go

```diff
@@ -0,0 +1,241 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package dockerfile
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+	"unicode"
+)
+
+// instructionName identifies the name of the instruction.
+type instructionName int
+
+const (
+	instrErr         instructionName = iota // an error occurred while scanning.
+	instrHealthCheck                        // a HEALTHCHECK instruction.
+	instrExpose                             // an EXPOSE instruction.
+	instrEOF                                // done scanning.
+)
+
+const (
+	exposeInstrMarker     = "expose "      // start of an EXPOSE instruction.
+	heathCheckInstrMarker = "healthcheck " // start of a HEALTHCHECK instruction.
+)
+
+var (
+	lineContinuationMarkers = []string{"`", "\\"} // denotes that the instruction continues to the next line.
+
+	instrMarkers = map[instructionName]string{ // lookup table for how an instruction starts.
+		instrExpose:      exposeInstrMarker,
+		instrHealthCheck: heathCheckInstrMarker,
+	}
+)
+
+// An instruction part of a Dockerfile.
+// Dockerfiles are of the following format:
+// ```
+// # Comment
+// INSTRUCTION arguments
+// ```
+type instruction struct {
+	name instructionName // the type of the instruction.
+	args string          // the arguments of an instruction.
+	line int             // line number at the start of this instruction.
+}
+
+// lexer holds the state of the scanner.
+type lexer struct {
+	name    string         // name of the Dockerfile; used only for error reports.
+	scanner *bufio.Scanner // line-by-line scanner of the contents of the Dockerfile.
+
+	curLineCount int              // line number scanned so far.
+	curLine      string           // current line scanned.
+	curArgs      *strings.Builder // accumulated arguments for an instruction.
+
+	instructions chan instruction //channel of discovered instructions.
+}
+
+// lex returns a running lexer that scans the Dockerfile.
+// The lexing logic is heavily inspired by:
+//   https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/text/template/parse/lex.go
+func lex(name string, reader io.Reader) *lexer {
+	l := &lexer{
+		name:         name,
+		scanner:      bufio.NewScanner(reader),
+		curArgs:      new(strings.Builder),
+		instructions: make(chan instruction),
+	}
+	go l.run()
+	return l
+}
+
+// next returns the next scanned instruction.
+func (lex *lexer) next() instruction {
+	return <-lex.instructions
+}
+
+// readLine loads the next line in the Dockerfile.
+// If we reached the end of the file, then isEOF is set to true.
+// If any unexpected error occurs during scanning, then err is not nil.
+func (lex *lexer) readLine() (isEOF bool, err error) {
+	if ok := lex.scanner.Scan(); !ok {
+		if err := lex.scanner.Err(); err != nil {
+			return false, fmt.Errorf("scan Dockerfile %s: %w", lex.name, err)
```

Comment:
I wonder if the lex's error cares about the file name information 🤔  If `lex` is just scanning and processing the content, maybe the file name information would be better suited in the error that wraps this error, that is in the caller error? Just a thought 💭

Reply:
Good point! removed